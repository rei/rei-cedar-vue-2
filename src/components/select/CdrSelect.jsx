import clsx from 'clsx';
import toArray from 'lodash-es/toArray';
import IconCaretDown from '../icon/comps/caret-down';
import size from '../../mixins/size';
import space from '../../mixins/space';
import style from './styles/CdrSelect.scss';

export default {
  name: 'CdrSelect',
  components: {
    IconCaretDown,
  },
  mixins: [size, space],
  inheritAttrs: false,
  model: {
    prop: 'value',
    event: 'change',
  },
  props: {
    /**
     * `id` for the select that is mapped to the label `for` attribute. If one is not provided, it will be generated.
    */
    id: String,
    /**
     * Label text. This is required for a11y even if hiding the label with `hideLabel`.
    */
    label: {
      type: String,
      required: true,
    },
    /**
     * Removes the label element but sets the select `aria-label` to `label` text for a11y.
    */
    hideLabel: Boolean,
    /**
     * Adds an option that is disabled and selected by default to serve as a `placeholder` for the select.
    */
    prompt: String,
    /**
     * Build options programatically with data. Array of objects [{ text: String, value: String}] to give greater control. Array of strings ['String'] for simpler setup (value and text will be the same).
    */
    options: {
      type: Array,
    },
    /** @ignore */
    value: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
    },
    /** @ignore */
    disabled: Boolean,
    /** @ignore */
    required: Boolean,
    /** DEPRECATED */
    multiple: Boolean,
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    // Use given id or generate one
    selectId() {
      return this.id ? this.id : this._uid; // eslint-disable-line no-underscore-dangle
    },
    baseClass() {
      return 'cdr-select';
    },
    selectClass() {
      return {
        [this.style['cdr-select']]: true,
        [this.style['cdr-select__prompt']]: !this.value,
        [this.style['cdr-select--multiple']]: this.multiple,
      };
    },
    labelClass() {
      return {
        [this.style['cdr-select__label']]: true,
        [this.style['cdr-select__label--disabled']]: this.disabled,
      };
    },
    caretClass() {
      return {
        [this.style['cdr-select__caret']]: true,
        [this.style['cdr-select__caret--disabled']]: this.disabled,
      };
    },
    selectWrapClass() {
      return {
        [this.style['cdr-select-wrap']]: true,
      };
    },
    inputListeners() {
      // https://vuejs.org/v2/guide/components-custom-events.html#Binding-Native-Events-to-Components
      // handles conflict between v-model and v-on="$listeners"
      const vm = this;
      return Object.assign(
        {},
        this.$listeners,
        {
          change(event) {
            if (vm.multiple) {
              const optArr = toArray(event.target.options);
              const selected = optArr.filter(o => o.selected === true).map(o => o.value);

              vm.$emit('change', selected, event);
            } else {
              vm.$emit('change', event.target.value, event);
            }
          },
        },
      );
    },
    selectEl() {
      return (
        <select
          class={clsx(this.sizeClass, this.selectClass)}
          id={this.selectId}
          multiple={this.multiple}
          disabled={this.disabled}
          required={this.required}
          aria-label={this.hideLabel ? this.label : null}
          ref="select"
          {...{ attrs: this.$attrs, on: this.inputListeners }}
          vModel={this.value}
        >

          {this.prompt
            && <option
              class="cdr-select__prompt"
              value=""
              disabled
              ref="prompt"
            >
              { this.prompt }
            </option>
          }
          {this.computedOpts.map(option => (
              <option
                key={option.text}
                value={option.value}
              >
                { option.text }
              </option>
          ))}
          {this.$slots.default}
        </select>
      );
    },
    labelEl() {
      const requiredEl = this.required ? (
        <span
          class={this.style['cdr-select__required-label']}
        >
          Required
        </span>
      ) : '';

      return !this.hideLabel ? (
        <label
          class={this.labelClass}
          for={this.selectId}
          ref="label"
        >{ this.label }
          {' '}
          {requiredEl}
        </label>
      ) : '';
    },
    infoEl() {
      return this.$slots.info ? (
        <span
          class={this.style['cdr-select__info-container']}
        >
          {this.$slots.info}
        </span>
      ) : '';
    },
    helperEl() {
      return this.$slots['helper-text'] ? (
        <span
          class={this.style['cdr-select__helper-text']}
        >
          {this.$slots['helper-text']}
        </span>
      ) : '';
    },
    computedOpts() {
      const optsArr = [];
      if (this.options) {
        this.options.forEach((o) => {
          const optObj = {};
          let text = '';
          let val = '';
          if (typeof o === 'string') {
            text = o;
            val = o;
          } else {
            const { text: t, value: v } = o;
            text = t;
            val = v;
          }
          optObj.text = text;
          optObj.value = val;
          optsArr.push(optObj);
        });
      }
      return optsArr;
    },
  },
  render() {
    return (
      <div class={clsx(this.space)}>
        {this.labelEl}
        {this.infoEl}
        <div class={this.selectWrapClass}>
          {this.selectEl}
          <icon-caret-down
          class={this.caretClass}
          />
        </div>
        {this.helperEl}
      </div>
    );
  },
};
