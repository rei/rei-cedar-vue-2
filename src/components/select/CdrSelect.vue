<template>
  <div :class="$style['cdr-input-group']">
    <label
      v-if="!hideLabel"
      :class="labelClass"
      :for="selectId"
      ref="label"
    >{{ label }}<span v-if="required">*</span></label>
    <select
      :class="selectClass"
      v-bind="$attrs"
      :id="selectId"
      :size="size"
      @change="onChange"
      ref="select"
      v-model="val"
      :required="required"
      :multiple="multiple"
      :aria-label="hideLabel ? label : null"
    >
      <option
        v-if="prompt"
        value=""
        disabled
        :hidden="!multiple"
        ref="prompt"
      >{{ prompt }}</option>
      <option
        v-for="option in computedOpts"
        v-if="option"
        :key="option.text"
        :value="option.value"
      >{{ option.text }}</option>
      <slot/>
    </select>
  </div>
</template>

<script>
import toArray from 'lodash/toArray';

/**
 * Cedar 2 component for select
 * **NOTE:** `v-model` is required.
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrSelect',
  inheritAttrs: false,
  model: {
    prop: 'extVal',
    event: 'change',
  },
  props: {
    /**
     * Label text.
    */
    label: {
      type: String,
      required: true,
    },
    /**
     * `id` for the select that is mapped to the label `for` attribute. If one is not provided, it will be generated.
    */
    id: String,
    /**
     * Removes the label element but sets the select `aria-label` to `label` text for a11y.
    */
    hideLabel: Boolean,
    /**
     * Adds an option that is disabled and selected by default to serve as a `placeholder` for the select.
    */
    prompt: String,
    /**
     * Build options programatically with data. Array of objects [{ text: String, value: String}] to give greater control. Array of strings ['String'] for simpler setup (value and text will be the same).
    */
    options: {
      type: Array,
    },
    /** @ignore */
    extVal: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      required: false,
    },
    /** @ignore */
    required: Boolean,
    /** @ignore */
    size: String,
    /** @ignore */
    multiple: Boolean,
  },
  data() {
    return {
      val: this.extVal,
    };
  },
  computed: {
    // Use given id or generate one
    selectId() {
      return this.id ? this.id : this._uid; // eslint-disable-line no-underscore-dangle
    },
    selectClass() {
      return {
        [this.$style['cdr-select']]: true,
        [this.$style['cdr-select--size']]: parseInt(this.size, 10) > 0,
      };
    },
    labelClass() {
      return {
        [this.$style['cdr-select__label']]: true,
        [this.$style['cdr-select__label--disabled']]: this.disabled,
      };
    },
    computedOpts() {
      const optsArr = [];
      if (this.options) {
        this.options.forEach((o) => {
          const optObj = {};
          let text = '';
          let val = '';
          if (typeof o === 'string') {
            text = o;
            val = o;
          } else {
            const { text: t, value: v } = o;
            text = t;
            val = v;
          }
          optObj.text = text;
          optObj.value = val;
          optsArr.push(optObj);
        });
      }
      return optsArr;
    },
  },
  watch: {
    extVal() {
      if (!this.multiple) {
        this.val = this.extVal;
      }
    },
  },
  mounted() {
    // initialize options as selected if multiple
    if (this.multiple) {
      const opts = toArray(this.$refs.select.options);
      opts.forEach((opt) => {
        const o = opt;
        if (this.val.indexOf(o.value) !== -1) {
          o.selected = true;
        }
      });
    }
  },
  methods: {
    onChange(e) {
      /**
       * Current input value. Fires when
       * @event input
       * @type string|array
       */
      if (this.multiple) {
        const optArr = toArray(e.target.options);
        const selected = optArr.filter(o => o.selected === true).map(o => o.value);
        this.val = selected;
        this.$emit('change', selected, e);
      } else {
        this.val = e.target.value;
        this.$emit('change', e.target.value, e);
      }
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/vars/CdrSelect.vars.pcss';
  @import './styles/CdrSelect.pcss';
</style>
