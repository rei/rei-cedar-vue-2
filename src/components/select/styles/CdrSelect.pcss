/* ==========================================================================
  # CdrSelect
  
  All values should map to variables in
  vars/CdrSelect.vars.pcss
  in order to allow for theming

  TOC:

    :Base - Select
      :States
========================================================================== */

.cdr-select {
  &.cdr-select {
    box-sizing: border-box;
    font-size: $select-font-size;
    line-height: $select-line-height;
    text-align: $select-text-align;
    color: $select-color;
    font-style: $select-font-style;
    font-weight: $select-font-weight;
    font-family: $select-font-family;
    width: 100%;
    display: block;
    border: $select-border;
    border-radius: $select-border-radius;
    height: $select-height;
    background-color: $select-background-color;
    text-transform: none;
    margin: 0;
  }

  &[multiple] {
    padding: 0;
  }

  & option {
    line-height: 1;
    box-sizing: border-box;
  }

  /* States
    ---------- */

  &[disabled] {
    background-color: $select-disabled-background;
  }

  /* Style variants
    ========================================================================== */

  &--size&[size] {
    height: auto;
  }

  &__label {
    @include maple-utility-20;

    display: block;
    margin: 0;
  }
}
