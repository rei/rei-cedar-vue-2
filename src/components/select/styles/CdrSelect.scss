@import '../../../css/settings/index.scss';
@import './vars/CdrSelect.vars.scss';
/* ==========================================================================
  # CdrSelect

  All values should map to variables in
  vars/CdrSelect.vars.pcss
  in order to allow for theming

  TOC:

  :Base - Select
  :States

  /* Style variants
========================================================================== */

.cdr-select {
  @include cdr-select-base-mixin;

  &__required-label {
    @include cdr-select-required-label-mixin;
  }

  &__label {
    @include cdr-select-base-label-mixin;

    &--large {
      @include cdr-select-large-mixin;
    }

    &--disabled {
      @include cdr-select-base-label-disabled-mixin;
    }
  }

  &__caret {
    @include cdr-select-base-caret-mixin;

    &--disabled {
      @include cdr-select-base-caret-disabled-mixin;
    }
  }

  &__info-container {
    @include cdr-select-info-container-mixin;
  }

  &__helper-text {
    @include cdr-select-helper-text-mixin;
  }

  // DEPRICATED STYLE
  // ========================================================================== */
  &--multiple {
    height: auto;
  }
  // ========================================================================== */

  /* @xs
  0px - 767px
  ========== */
  @include xs-mq-only {
    &--large\@xs {
      @include cdr-select-large-mixin;
    }
  }

    /* Size variants
    ========================================================================== */

    &--large {
      @include cdr-select-large-mixin;
    }

  /* @sm
    768px - 991px
    ========== */
  @include sm-mq-only {
    &--large\@sm {
      @include cdr-select-large-mixin;
    }
  }

  /* @md
    992px - 1199px
    ========== */
  @include md-mq-only {
    &--large\@md {
      @include cdr-select-large-mixin;
    }
  }

  /* @lg
    1200px and up
    ========== */

  @include lg-mq-only {
    &--large\@lg {
      @include cdr-select-large-mixin;
    }
  }
}

/* ==========================================================================
   # SELECT WRAPPER
   ========================================================================== */

/* Sets new origin for relative positioning of Select and Caret */
.cdr-select-wrap {
  position: relative;
}
