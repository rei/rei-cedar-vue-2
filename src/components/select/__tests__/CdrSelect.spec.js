import { shallowMount, mount } from '@vue/test-utils';
import CdrSelect from 'componentdir/select/CdrSelect';

describe('cdrSelect', () => {
  it('renders correctly', () => {
    const wrapper = mount(CdrSelect, {
      propsData: {
        label: 'Label Test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders required label correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'Label Test',
        required: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('hide-label sets aria-label correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
        hideLabel: true,
      },
    });
    expect(wrapper.vm.$refs.select.hasAttribute('aria-label', 'test')).toBe(true);
  });

  it('maps select id to label for correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'testing',
        id: 'test',
      },
    });
    expect(wrapper.vm.$refs.select.id).toBe(wrapper.vm.$refs.label.htmlFor);
  });

  it('generates an id correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'testing',
      },
    });
    expect(wrapper.vm.$refs.select.id).toBe(wrapper.vm._uid.toString());
  });

  it('sets select name attribute correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'Label Test',
      },
      attrs: {
        name: 'testing',
      }
    });
    expect(wrapper.vm.$refs.select.hasAttribute('name', 'testing')).toBe(true);
  });

  it('sets select disabled attribute correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
      },
      attrs: {
        disabled: true,
      }
    });
    expect(wrapper.vm.$refs.select.hasAttribute('disabled')).toBe(true);
  });

  it('sets select required attribute correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
        required: true,
      },
    });
    expect(wrapper.vm.$refs.select.hasAttribute('required', 'required')).toBe(true);
  });

  it('sets select autofocus attribute correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
      },
      attrs: {
        autofocus: true,
      }
    });
    expect(wrapper.vm.$refs.select.hasAttribute('autofocus')).toBe(true);
  });

  it('sets select multiple attribute correctly', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'Label Test',
        multiple: true,
        value: [],
      },
    });
    expect(wrapper.vm.$refs.select.hasAttribute('multiple')).toBe(true);
  });

  it('emits input event with correct value', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
        value: '2',
        options: ['1', '2'],
      },
    });
    const select = wrapper.find({ ref: 'select'});
    wrapper.setProps({ value: '1' });
    select.trigger('input');
    expect(wrapper.emitted().input[0][0]).toBe('1');
  });

  // Deprecated event and should be removed
  // when the change event is removed.
  it('emits change event with correct value', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
        value: '4',
        options: ['3', '4'],
      },
    });
    const select = wrapper.find({ ref: 'select'});
    wrapper.setProps({ value: '3' });
    select.trigger('change');
    expect(wrapper.emitted().change[0][0]).toBe('3');
  });

  it('emits input event with correct value for multiple', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
        multiple: true,
        value: ['1', '2'],
        options: [{
          value: '1',
          text: 'one',
        },
        {
          value: '2',
          text: 'two',
        },
        {
          value: '3',
          text: 'three',
        }],
      },
    });
    wrapper.setProps({ value: ['1', '3'] });
    const propValues = wrapper.vm.value;
    for(let o of wrapper.vm.$refs.select.options) {
      propValues.indexOf(o.value) >= 0 ? o.selected = true : o.selected = false;
    }
    const select = wrapper.find({ ref: 'select'});
    select.trigger('input');
    expect(wrapper.emitted().input[0][0]).toEqual(['1', '3']);
  });

  // Deprecated event and should be removed
  // when the change event is removed.
  it('emits change event with correct value for multiple', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
        multiple: true,
        value: ['1', '2'],
        options: [{
          value: '1',
          text: 'one',
        },
        {
          value: '2',
          text: 'two',
        },
        {
          value: '3',
          text: 'three',
        }],
      },
    });
    wrapper.setProps({ value: ['1', '3'] });
    const propValues = wrapper.vm.value;
    for(let o of wrapper.vm.$refs.select.options) {
      propValues.indexOf(o.value) >= 0 ? o.selected = true : o.selected = false;
    }
    const select = wrapper.find({ ref: 'select'});
    select.trigger('input');
    expect(wrapper.emitted().change[0][0]).toEqual(['1', '3']);
  });

  // NOTE - can't use v-model directly here, targeting the `data` prop instead
  it('updating v-model data updates the select', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
        value: '4',
        options: ['3', '4'],
      },
    });
    const select = wrapper.find({ ref: 'select' });
    wrapper.setProps({value: '3'});
    expect(select.element.value).toBe('3');
  });

  it('renders info slot', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
      },
      slots: {
        'info': 'ðŸ¤ ',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders helper-text slot', () => {
    const wrapper = shallowMount(CdrSelect, {
      propsData: {
        label: 'test',
      },
      slots: {
        'helper-text': 'very helpful',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
