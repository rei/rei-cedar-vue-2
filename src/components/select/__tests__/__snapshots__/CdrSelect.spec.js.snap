// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdrSelect renders correctly 1`] = `
<div
  class=""
>
  <label
    class="cdr-select__label"
    for="1"
  >
    Label Test 
  </label>
  <div
    class="cdr-select-wrap"
  >
    <select
      class="cdr-select  cdr-select cdr-select__prompt"
      id="1"
    />
    <svg
      class="cdr-icon  cdr-select__caret"
      role="presentation"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 16c.273 0 .521-.11.702-.288l5.005-5.005a1 1 0 0 0-1.414-1.414L12 13.586 7.705 9.29a1 1 0 0 0-1.412 1.417l4.98 4.98c.182.193.44.313.727.313z"
      />
    </svg>
  </div>
</div>
`;

exports[`cdrSelect renders helper-text slot 1`] = `
<div
  class=""
>
  <label
    class="cdr-select__label"
    for="50"
  >
    test 
  </label>
  <div
    class="cdr-select-wrap"
  >
    <select
      class="cdr-select  cdr-select cdr-select__prompt"
      id="50"
    />
    <icon-caret-down-stub
      class="cdr-select__caret"
      size="medium"
      space=""
    />
  </div>
  <span
    class="cdr-select__helper-text"
  >
    very helpful
  </span>
</div>
`;

exports[`cdrSelect renders info slot 1`] = `
<div
  class=""
>
  <label
    class="cdr-select__label"
    for="47"
  >
    test 
  </label>
  <span
    class="cdr-select__info-container"
  >
    ðŸ¤ 
  </span>
  <div
    class="cdr-select-wrap"
  >
    <select
      class="cdr-select  cdr-select cdr-select__prompt"
      id="47"
    />
    <icon-caret-down-stub
      class="cdr-select__caret"
      size="medium"
      space=""
    />
  </div>
</div>
`;

exports[`cdrSelect renders required label correctly 1`] = `
<div
  class=""
>
  <label
    class="cdr-select__label"
    for="5"
  >
    Label Test 
    <span
      class="cdr-select__required-label"
    >
      Required
    </span>
  </label>
  <div
    class="cdr-select-wrap"
  >
    <select
      class="cdr-select  cdr-select cdr-select__prompt"
      id="5"
      required="required"
    />
    <icon-caret-down-stub
      class="cdr-select__caret"
      size="medium"
      space=""
    />
  </div>
</div>
`;
